src/chain_service/faucet.ts

```diff
@@ -0,0 +1,201 @@
+import { ApiPromise, WsProvider, Keyring } from "@polkadot/api";
+import { checkAddress, cryptoWaitReady } from "@polkadot/util-crypto";
+import BN from "bn.js";
+import { ApiManager, API } from "./api.js";
+import { Config } from "../config.js";
+import {
+  AddressError,
+  DispatchError,
+  NoFundsError,
+  RateError,
+} from "./faucetErrors.js";
+import { SlackNotifier } from "../slack_service/slack.js";
+
+export default class Faucet {
+  private config: Config;
+  private apiManager: ApiManager;
+  private fundMap: Map<string, number[]>;
+  private slackNotifier: SlackNotifier;
+
+  constructor(
+    config: Config,
+    slackNotifier: SlackNotifier,
+    apiManager: ApiManager,
+  ) {
+    this.config = config;
+    this.apiManager = apiManager;
+    this.slackNotifier = slackNotifier;
+    this.fundMap = new Map<string, number[]>();
+  }
+
+  async init() {
+    await this.apiManager.getApi();
+  }
+
+  async send(address: string) {
+    let api = await this.apiManager.getApi();
+    return new Promise<string>(async (resolve, reject) => {
+      // Waiting for crypto library to be ready
+      await cryptoWaitReady();
+
+      // Reject early if the address has an incorrect format
+      // or limit has been hit
+      const check = checkAddress(address, this.config.getaddressPrefix());
+      if (!check[0]) {
+        return reject(
+          new AddressError(
+            `Invalid address! The expected SS58 address prefix used for encoding ${this.config.getaddressPrefix()}, visit https://github.com/paritytech/substrate/blob/e232d78dd5bafa3bbaae9ac9db08f99e238392db/primitives/core/src/crypto.rs#L444 for reference`,
+            address,
+          ),
+        );
+      }
+      // Handle limits
+      let remainingTime = this.isAddressAllowed(address);
+      if (remainingTime) {
+        return reject(
+          new RateError(`Address ${address} has reached limit`, remainingTime),
+        );
+      }
+      this.addFundEvent(address);
+
+      // Derive sender object from secret
+      const keyring = new Keyring({ type: "sr25519" });
+      keyring.setSS58Format(this.config.getaddressPrefix());
+
+      let sender;
+      try {
+        sender = keyring.addFromUri(this.config.getSecret());
+      } catch {
+        return reject(`Service Error: Invalid Mnemonic`);
+      }
+
+      const nonce = await api.api.rpc.system.accountNextIndex(sender.publicKey);
```

Comment:
Are we actually able to support multiple incoming faucet requests by different users or would this nonce here be a dealbreaker? Maybe it works perfectly fine, just wondering

Reply:
I tested it by sending multiple requests at the "same time" (manually) and they all worked, or at least entered in the same block.
What is attempted to do here by fetching the nonce is exactly that, to allow multiple fund events simultaneously. 

Now I wonder if the nonce is able to increase if the transactions arrive faster that what could manually be tested. I guess that to be on the safe side, we could implement a mutex similar to what we did [here](https://github.com/pendulum-chain/spacewalk-testing-service/blob/main/src/vault_service/vault.ts#L43). The problem I see with something like that is that it could delay a lot the completion. WDYT?

Reply:
Ah nice, I was not fully aware how the `accountNextIndex` works. But since this apparently is taking the pending transactions of the transaction pool of the node into account, I think it's fine and should indeed work as expected. 

I already noticed that the Spacewalk testing service was able to submit multiple extrinsics in the same block so I can confirm that this works. Let's not add a mutex as it's not necessary then.

Reply:
Yes I think it is also important for that to keep the `await` in cases like in the example to ensure that the transaction reached the pool. 
It is still not clear to me what would happen if 2 transactions were to be sent in which the first transaction does not reach the pool before the second one is sent.